From 4b2e669441438a6ae91c82a24c796894aa6a53b2 Mon Sep 17 00:00:00 2001
From: Alejandro Colomar <alx@kernel.org>
Date: Thu, 15 Jan 2026 01:14:46 +0100
Subject: [PATCH 1/3] lib/chkhash.c: is_valid_hash(): Update comment

This checks the entire shadow(5) 2nd field, which is more than just
a hash.

Reported-by: Tobias Stoeckmann <tobias@stoeckmann.org>
Signed-off-by: Alejandro Colomar <alx@kernel.org>
(cherry picked from commit bce404a7a831de911cf20fc8d4c548957f70bc72)
---
 lib/chkhash.c | 4 +---
 1 file changed, 1 insertion(+), 3 deletions(-)

diff --git a/lib/chkhash.c b/lib/chkhash.c
index 348dfe7478f8..48aaa417e9a9 100644
--- a/lib/chkhash.c
+++ b/lib/chkhash.c
@@ -31,9 +31,7 @@ match_regex(const char *pattern, const char *string)
 
 
 /*
- * is_valid_hash - check if the given string is a valid password hash
- *
- * Returns true if the string appears to be a valid hash, false otherwise.
+ * is_valid_hash - check if the string is a valid shadow(5) 2nd field.
  *
  * regex from: https://man.archlinux.org/man/crypt.5.en
  */
-- 
2.53.0


From b72386978e0c994c5efc91d548da68a8d97088ab Mon Sep 17 00:00:00 2001
From: Alejandro Colomar <alx@kernel.org>
Date: Thu, 15 Jan 2026 01:24:10 +0100
Subject: [PATCH 2/3] lib/chkhash.c: is_valid_hash(): Comment meaning of !hash
 and *

Signed-off-by: Alejandro Colomar <alx@kernel.org>
(cherry picked from commit 9a86c515a1e7caeeb9e60fa63871170a2dfd98f0)
---
 lib/chkhash.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/lib/chkhash.c b/lib/chkhash.c
index 48aaa417e9a9..9123038adef0 100644
--- a/lib/chkhash.c
+++ b/lib/chkhash.c
@@ -38,12 +38,14 @@ match_regex(const char *pattern, const char *string)
 bool 
 is_valid_hash(const char *hash) 
 {
+	// Password temporarily locked
 	hash = strprefix(hash, "!") ?: hash;
 
 	// Passwordless account; discouraged
 	if (streq(hash, ""))
 		return true;
 
+	// Password permanently locked (and forgotten)
 	if (streq(hash, "*"))
 		return true;
 
-- 
2.53.0


From 4486e3e0f8b74e6af27b5dcb4114fd776afdd605 Mon Sep 17 00:00:00 2001
From: Alejandro Colomar <alx@kernel.org>
Date: Fri, 16 Jan 2026 01:52:32 +0100
Subject: [PATCH 3/3] man/shadow.5.xml: Document "*"

Signed-off-by: Alejandro Colomar <alx@kernel.org>
(cherry picked from commit 6be13b2f84a2c1a0d0f4129b5258b4b443e7f86c)
---
 man/shadow.5.xml | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/man/shadow.5.xml b/man/shadow.5.xml
index dc7fdfde542f..d3cc03982c47 100644
--- a/man/shadow.5.xml
+++ b/man/shadow.5.xml
@@ -82,8 +82,12 @@
 	    If the password field begins with an exclamation mark <emphasis>!</emphasis>,
 	    the password is locked.
 	    The remaining characters on the
-	    line represent the password field before the password was
-	    locked.
+	    line represent the password hash.
+	  </para>
+	  <para>
+	    If the password hash consists of a star <emphasis>*</emphasis>,
+	    password access is effectively disallowed;
+	    this is because no password can produce a hash like this.
 	  </para>
 	  <para>
 	    Refer to <citerefentry><refentrytitle>crypt</refentrytitle>
-- 
2.53.0

